src/Tests/Microsoft.DotNet.ApiCompat.Tests/RegexStringTransformerTests.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.DotNet.ApiCompat;
+using Xunit;
+
+namespace Microsoft.DotNet.ApiCompatibility.Tests
+{
+    public class RegexStringTransformerTests
+    {
+        [Fact]
+        public void Transform_CaptureGroupPatternDoesNotMatchInput_ReturnsInput()
+        {
+            const string CaptureGroupPattern = "(abc)def";
+            const string ReplacementPattern = "$1";
+            const string Input = "ghi";
+
+            string output = new RegexStringTransformer(CaptureGroupPattern, ReplacementPattern).Transform(Input);
+
+            Assert.Equal(Input, output);
+        }
+
+        [Fact]
+        public void Transform_ReplacementPatternWithoutCaptureGroups_ReturnsReplacementPattern()
+        {
+            const string CaptureGroupPattern = "(abc)d*";
+            const string ReplacementPattern = "xyz";
+            const string Input = "abc";
+
+            string output = new RegexStringTransformer(CaptureGroupPattern, ReplacementPattern).Transform(Input);
+
+            Assert.Equal(ReplacementPattern, output);
+        }
+
+        [Fact]
+        public void Transform_ReplacementPatternWithTooManyReplacementMarkers_ReturnOutputWithoutTransformedReplacementMarkers()
+        {
+            const string CaptureGroupPattern = "(abc)(def)ghi";
+            const string ReplacementPattern = "1:$1, 2:$2, 3:$3";
+            const string Input = "abcdefghi";
+
+            string output = new RegexStringTransformer(CaptureGroupPattern, ReplacementPattern).Transform(Input);
+
+            Assert.Equal("1:abc, 2:def, 3:$3", output);
+        }
+
+        [Fact]
+        public void Transform_SameNumberOfGroupsAndMarkers_ReturnsExpected()
+        {
+            const string CaptureGroupPattern = @".+\\(.+)\\(.+)";
+            const string ReplacementPattern = "lib/$1/$2";
+            const string Input = @"C:\git\runtime\artifacts\bin\System.Linq\Debug\net7.0-android\System.Linq.dll";
+
+            string output = new RegexStringTransformer(CaptureGroupPattern, ReplacementPattern).Transform(Input);
+
+            Assert.Equal("lib/net7.0-android/System.Linq.dll", output);
+        }
+
+        [Fact]
+        public void Transform_MultiplePatterns_ReturnsExpected()
```

Comment:
I'd also add a test that forces catastrophic backtracking just to make sure that the engine succesfully times out. You can try pattern `^(\w+\s?)*$` with an expression like `An input string that takes a very very very very very very very very very very very long time!`